hybrid monte carlo

draneyj opened the pull request:
**Summary**

This pull request adds hybrid/Hamiltonian Monte Carlo for atomic and rigid/small systems. It's based on the work done by @ajsilveira as described in the related issue. We did this for a project in a grad student software engineering for computer science class at Princeton, which just finished. It's in a working and sensible form but we'd like some guidance on how to cross the finish line.

**Related Issue(s)**

closes #565: [Feature Request] hybrid monte carlo

**Author(s)**

Authors from the original issue in 2017:
Charlles R. A. Abreu (abreu@eq.ufrj.br)
Ana J. Silveira (asilveira@plapiqui.edu.ar)

Updated the code this semester, 2023:
Jack S. Draney (jacksdraney@gmail.com) <-- corresponding
Louis E. S. Hoffenberg (lhoff@princeton.edu)
Baris E. Ugur (bu9134@princeton.edu)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Changes in the pull request will not break backward compatibility for inputs

**Implementation Notes**

Fix::FixHMC was added to the MC package. States are stored/restored by manually saving and altering many values, but also by looping over the peratom vector of the Atom class to make sure we got everything. The only code outside of this that was altered was in rigid/small, where we made some things public so the FixHMC could change the values. We could, alternatively, friend FixHMC or derive a new rigid/small that just adds getter and setter functions for the appropriate values so FixHMC can alter them.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

We are somewhat new to contributing to open source so please be patient as we figure out how to do stuff on github we've never done before, etc. I've read all the docs pages on contributing, but it is a lot of information, so I apologize in advance for whatever I miss along the way.

draneyj committed:
adding source files from original implementation

draneyj committed:
hacked away at the outdated code until it compiled

draneyj committed:
now runs without error

draneyj committed:
properly initialize vector output

draneyj committed:
trying to store peratom data

draneyj committed:
added restoration function

draneyj committed:
re-added restoration function

draneyj committed:
restored accidentally deleted code

draneyj committed:
runs with new atom copy

draneyj committed:
peratom copy working (for x at least)

draneyj committed:
remove comments in restore_saved_state

draneyj committed:
removed warning aboout rigid support

draneyj committed:
added templates to class, allowing error

draneyj committed:
Merge pull request #1 from draneyj/pr_practice

added templates to class, allowing error

draneyj committed:
missed a line I needed to remove

draneyj committed:
add body copy for rigid bodies

draneyj committed:
random velocities for bodies and mom arg for zero vcm

draneyj committed:
modernize arg parse syntax

draneyj committed:
properly weight vcm for rigid bodies

lh0534 committed:
Added group filter to body random velocities

baris-eser-ugur committed:
Added fix_hmc.rst

baris-eser-ugur committed:
Syntax, Examples and Description added to fix_hmc.rst

baris-eser-ugur committed:
Documentation description modified, related commands, default and references added.

draneyj committed:
added more keywords to the documentation

draneyj committed:
trying to accept chaning number of atoms

draneyj committed:
remove erroneous line

draneyj committed:
initialize map flag

draneyj committed:
comment out rot keyword

draneyj committed:
added atom tag storage

lh0534 committed:
Updated GA CI files to work on draneyj/lammps_hmc repo.

draneyj committed:
added xcmimage to restoration, fixed lost atoms

draneyj committed:
Merge branch 'develop' of github.com:draneyj/lammps_hmc into develop

draneyj committed:
adding more stored vectors just in case

draneyj committed:
temporarily removing parameter tuning

lh0534 committed:
Touched up the docs file fix_hmc.rst

lh0534 committed:
Added the default to the docs file fix_hmc.rst

lh0534 committed:
Clarified the hmc algorithm in the doc file fix_hmc.rst

draneyj committed:
remove tuning

draneyj committed:
removed random p when move accepted

draneyj committed:
remove unnecessary code and comments

draneyj committed:
uncomment reverse_comm

draneyj committed:
check for nullptr at source before copying or allocating

draneyj committed:
make resample on accept an option

draneyj committed:
make resample on accept default no

draneyj committed:
current_peratom updated every time

draneyj committed:
minor edit to reset atom_map

draneyj committed:
added memory usage calculation

draneyj committed:
formatting

draneyj committed:
more formatting

draneyj committed:
add contributing authors

draneyj committed:
moving hmc to MC package

lh0534 committed:
Added ra (resampling upon acceptance) to hmc doc).

lh0534 committed:
touch up docs

draneyj committed:
trying to increase speed by minimizing reallocations

draneyj committed:
faster by not reallocating peratom vector

draneyj committed:
removed realloc to test

draneyj committed:
only copying ntotal in save peratom

draneyj committed:
Merge branch 'norealloc' into develop

draneyj committed:
reallocation conflict resolved

draneyj committed:
fixing merge issues

draneyj committed:
fixing merge issues

lh0534 committed:
Added MC pkg to CI files

draneyj committed:
merge upstream

draneyj committed:
add line at eof

draneyj committed:
add set_v back to public

draneyj committed:
rename git blame ignore file

draneyj committed:
exclude .git-blame-ignore

draneyj committed:
change my email

akohlmey committed:
revert protection settings and add FixHMC as friend class instead

akohlmey committed:
make the Body struct standalone so we can use a forward declaration

akohlmey committed:
documentation intergration and fixes

akohlmey committed:
build system integration and handling of dependency on fix rigid/small

akohlmey committed:
update for coding convention compliance

akohlmey commented:
@draneyj thanks for your submission. I've made a few mostly cosmetic changes that properly integrate the added files into the build procedure and the manual build and indexing and so on. Also there are some coding style changes to make it pass "make check" in the src folder. Finally, I've looked at the access to the Body struct of fix rigid/small and moved it outside the class declaration and called it RigidSmallBody instead. With the friend class status for FixHMC this allows to drop the fix_rigid header include from the fix_hmc.h header (which is an important improvement).

There are a few fundamental things that I have comments on, but also there is significant potential to use a few more modern APIs to do some steps in a cleaner fashion. I will add comments to the individual parts of the code with some suggestions for how to address the issues over the next days.

akohlmey committed:
replace non-ASCII characters so that PDF builds won't fail

akohlmey committed:
explain dependency on RIGID package

akohlmey committed:
initialized all class member pointer to NULL

akohlmey committed:
small doc fix

akohlmey committed:
simplify creation of internal fixes and computes, use unique ID

akohlmey committed:
some more simplification and cleanup

akohlmey committed:
must move function to Fix::init() since queried info may not be current during constructor

akohlmey committed:
avoid memory leaks and uninitialized data

akohlmey committed:
plug some more memory leaks

akohlmey commented on src/MC/fix_hmc.cpp:
```diff
@@ -0,0 +1,1185 @@
+
+  // Initialize arrays for managing global virial terms:
+  nv = ne;
+  for (j = 0; j < modify->nfix; j++)
```
Accessing internal members of the `Modify` class should be avoided for anything but some select core classes. We have added new accessor APIs for those purposes (see issue #2870). So this loop would be replaced by:
```c++
for (const auto &ifix : modify->get_fix_list())
  if (ifix->virial_global_flag) nv++;
```
Same applies to the other loops below.

akohlmey commented on src/MC/fix_hmc.cpp:
```diff
@@ -0,0 +1,1185 @@
+  if (rigid_flag) first++;
+  for (int i = first; i < modify->nfix; i++)
+    if (modify->fix[i]->virial_peratom_flag || modify->fix[i]->virial_global_flag) {
+      if (comm->me == 0) printf("Fix %s defined after fix hmc.\n", modify->fix[i]->style);
```
Please don't use printf(), but `utils::logmesg()` for any kind of output, so the redirection to files for out works properly. The function supports printf() style formatted strings but uses python-like fmt::format() syntax with '{}' instead of '%s'.

akohlmey commented on src/MC/fix_hmc.cpp:
```diff
@@ -0,0 +1,1185 @@
+  int n = item + 1;
+  if (n == 1) {
+    double acc_frac = naccepts;
+    acc_frac /= MAX(1, nattempts);
```
This is redundant since it is already available as scalar compute.

akohlmey reviewed (changes requested):
I've already done a some formal changes but also some cleanup and modernization of some constructs, but there is still substantial amount of work left (e.g. all the loops over fixes and computes). Testing with valgind's memory checker expose some conceptional problem: the call to `setup_arrays_and_pointers()` is misplaced in the constructor and was leading to uninitialized data access. It has to come during `Fix::init()` instead, so that it checks up-to-date data before the run is started.

There should be multiple (small!) example cases added to the examples tree.
I've created a minimal test based on the melt example by replacing fix nve with:
`fix             1 all hmc 50 6234 1.5 flexible`. However, this leads to crashes. I haven't looked closer, but I suspect that it is because MC moves that are too disruptive for stable MD propagation are accepted.

There are likely more issues and other reviewers may request changes of their own, but let's get these more obvious issue sorted out first.

akohlmey commented:
> I've created a minimal test based on the melt example by replacing fix nve with:
> `fix 1 all hmc 50 6234 1.5 flexible`. However, this leads to crashes. I haven't looked closer, but I suspect that it is because MC moves that are too disruptive for stable MD propagation are accepted.

I've made some more experiments and it looks like the crashes are related to neighbor list rebuilds (which will migrate atoms between subdomains and also change their order in the local arrays due to that). If I turn neighbor list rebuilds off or do them with a fixed frequency that is an integer multiple of the frequency of the HMC moves, then no crashes happen and no "crazy" energies are observed.

akohlmey committed:
small programming style update

akohlmey committed:
check for tags and compatible neighbor list settings

draneyj committed:
remove access to internal members of modify in loops

draneyj committed:
removed redundant acc_frac calculation in compute_vector

draneyj committed:
change how check for additional virial fixes is done

draneyj committed:
initialize stored_body

draneyj committed:
check for rigid flag before fix_rigid

draneyj commented:
The crashing is strange. I'll look into it soon and see if I can figure out why things go haywire. With your updates, even with all the settings the same, my rigid example tests are no longer working also. If I remove any requirements on neighbor list build frequency, my flexible simulation still works. With the prescribed neighbor list modifications, it crashes. I'll upload my example scripts tomorrow.

akohlmey commented:
Here is one thought that may help to simplify things. There is currently a lot of programming effort included to store per-atom data. For that purpose an internal fix command (fix STORE/ATOM) exists. It is used in several places:
```sh
$ grep -l STORE/ATOM src/*.cpp src/*/*.cpp
src/balance.cpp
src/compute_chunk_atom.cpp
src/compute_displace_atom.cpp
src/compute_msd.cpp
src/compute_vacf.cpp
src/dump_custom.cpp
src/fix_adapt.cpp
src/fix_store_atom.cpp
src/variable.cpp
src/AMOEBA/pair_amoeba.cpp
src/CORESHELL/compute_temp_cs.cpp
src/EXTRA-COMPUTE/compute_hma.cpp
src/EXTRA-FIX/fix_nonaffine_displacement.cpp
src/EXTRA-MOLECULE/bond_harmonic_restrain.cpp
src/FEP/fix_adapt_fep.cpp
src/REPLICA/tad.cpp
src/VTK/dump_vtk.cpp
```
Since it is internal, you have to read the fix_store_atom.cpp and fix_store_atom.h sources and the use cases within LAMMPS to learn how it works.

draneyj committed:
properly force reneighbor on next timestep

draneyj commented:
> > I've created a minimal test based on the melt example by replacing fix nve with:
> > `fix 1 all hmc 50 6234 1.5 flexible`. However, this leads to crashes. I haven't looked closer, but I suspect that it is because MC moves that are too disruptive for stable MD propagation are accepted.
> 
> I've made some more experiments and it looks like the crashes are related to neighbor list rebuilds (which will migrate atoms between subdomains and also change their order in the local arrays due to that). If I turn neighbor list rebuilds off or do them with a fixed frequency that is an integer multiple of the frequency of the HMC moves, then no crashes happen and no "crazy" energies are observed.

I've corrected the method for forcing neighbor list rebuilds and I believe it alleviated some of these issues. Crazy energies observed with in.melt have become far more reasonable numbers. I've also removed the requirements for neighbor list settings as a result.

draneyj committed:
setup fix_rigid before saving state

draneyj committed:
add examples for hmc

draneyj committed:
changed vector output

akohlmey committed:
silence compiler warnings about unused variables

akohlmey committed:
update fix hmc examples and add reference log files

akohlmey commented:
@draneyj I've removed some unused variable declarations and looked at the examples. I removed TAB characters (TABs are evil!) and made some adjustments for more efficient running, e.g. by optimizing the neighbor list setting (for the flexible it was building neighbor lists in every step which is wasteful). The timestep for fix rigid was far too large with 1fs, changing it to a more reasonable 0.25fs avoids crashes. I've also recorded and committed reference output.

draneyj commented:
@akohlmey Oh no! I thought I had spaces instead of tabs set in my IDE! Thank you for fixing that.

Here's my plan moving forward: I'm going to keep storing and restoring the data tracked in atom->peratom as is. It automatically captures a lot of data without a lot of thought, which will hopefully minimize the amount of maintenance the code will require. That being said, there are many values not ttracked in atom->peratom that I need to track more manually. I'll go ahead and make some nice vectors out of these with the STORE/ATOM internal fix, for simplicity. Because I want fix hmc to be as widely usable as possible, and because I only know the little corner of MD and LAMMPS I use in my research (which doesn't actually use hybrid monte carlo), I'd appreciate your input on what per-atom values should be stored and restored at end_of_step, and which can be left to be recalculated based on the altered atomic data:

* atom tags, if present
* all the rigid body per-atom data if there is a rigid system
* energy (eatom) <-- signficant to restore at end_of_step? does anything use it before it's recalculated next step?
* virial (vatom) <-- signficant to restore at end_of_step? does anything use it before it's recalculated next step?
* others I am missing?

Other thoughts maybe you have an opinion on:
Currently the fix implements its own integration fix. Would it be sensible to have the user define their own and pass the fix id to the hmc fix as an argument? That way they can choose what kind of fix they want (small or nve or whatever). This would be added work, and I'm unsure if it's worth it.

If we disallow fix balance, we can assume domains remain the same, so we don't need to do any communication when restoring states if we also store and restore ghost atom data, since everything on the other processors should also be reset. If I take this route, maybe manually storing and restoring data could make more sense than STORE/ATOM which migrates data between processors.

akohlmey committed:
Merge branch 'develop' into draneyj/develop

sjplimp commented:
Hi - thanks for working on a new hybrid MC capability for LAMMPS.  I
read thru your new files and the discussion.  There are some things I
don't understand conceptually, and also code suggestions I have.  The
latter may be influenced by my limited understanding of the former.
I'll break my feedback into multiple comments for readability.

sjplimp commented:
(1) Doc page for fix_hmc command

This does not have enough context for the average LAMMPS user,
especially one not familiar with HMC.  E.g. why use it, what kinds of
systems is it good for.  What parameters affect its performance
(e.g. Nevery, timestep, Temp), how should they be set, what should be
monitored as it runs.  Also a basic description of how the algorithm
works.  See the doc pages for the hyper or prd commands which cover
most of this in the context of hyper and parallel replica dynamics,
including giving a pseudo-code outline of the algorithm.

I think HMC requires strict NVE simulations?  That should be
documented.  There are many LAMMPS commands which would break that and
not be easily detected.  E.g. any thermostat, fix move, fix deposit,
fix bond/react, etc.  Or any non-equilibrium MD construct. 

More specific Qs for doc page:
Change "ro" keyword to "resample".
What does it mean if group != all ?
What should the compute scale and compute vector outputs be for a successful simulation ?

sjplimp commented:
(2) Timestep for HMC runs:

I thought HMC was typically used with a larger-than-MD timestep so
that the dynamics has a better chance to relax the system to a smaller
energy?  Maybe that is a different flavor of HMC than this one?
Regardless, the timestep should be discussed in the doc page, both for rigid and
non-rigid systems.  The timestep in the rigid SPC/E example script is
1/4 fs, which seems tiny for rigid water?

sjplimp commented:
(2) Timestep for HMC runs:

I thought HMC was typically used with a larger-than-MD timestep so
that the dynamics has a better chance to relax the system to a smaller
energy?  Maybe that is a different flavor of HMC than this one?
Regardless, the timestehp should be discussed in the doc page, both for rigid and
non-rigid systems.  The timestep in the rigid SPC/E example script is
1/4 fs, which seems tiny for rigid water?

sjplimp commented:
(3) Fix_hmc inputs and setup():

Check that seed > 0, ditto for temp > 0.0
Lots of LAMMPS scripts perform successive runs.
So init() and setup() may be called multiple times.  
Some of their operations look like they are only meant to be done once.

sjplimp commented:
(4) Integrator(s):

I don't see why fix_hmc should define its own integration fix.  For
"flexible", I'm not seeing how the created fix is even stored?  For
the rigid body option, the ID of an existing fix rigid could be an
argument.  Several other LAMMPS commands do this.  Is fix hmc supposed
to work with a system of solvated rigid bodies?  Which is combination
of both flexible and rigid.  In which case the input script would typically
define both a fix nve and fix rigid/small for 2 parts of the system.

sjplimp commented:
(5) Timestep counter:

If an MC energy change is rejected, should the timestep counter be
reset to the beginning of the last store?

sjplimp commented:
(6) Neighbor list:

I saw some discussion about preventing neighbor list rebuilds during
each MD run.  This seems like a dangerous idea for a method,
which if I understand, relies on exact NVE accuracy for its MC
decisions to be correct.  I think the HMC method should "work
correctly" for any choice of Nevery and timestep.  And that the user
should ensure neighbor lists are built as often as needed during the
trial runs.  I would view it as the job of the save/restore logic to
correctly restore the simulation to its previous state, which can
leverage lots of existing LAMMPS functionality.  See the next comment.

sjplimp commented:
(7) Saving and restoring state:  (a long comment!)

The current save_current_state() and restore_saved_state() logic is
not suitable for allowing neighbor list rebuilds during a trial run.
I also don't think those methods should directly access per-atom or rigid body
data.

I suggest changing save_current_state() and restore_saved_state() as
follows for flexible systems:

(a) do not save ghost atom state
(b) invoke a new pack_state() method in AtomVec to return the "state" of a single owned atom
    as a vector of doubles
(c) the "state" is the set of quantities which vary during NVE dynamics.  
    For most atom styles this could be just x,v,image.  
    Also omega, angmom, dipole for select styles.  
    This would also allow styles that store Bonus
    data, like quaternions for ellipsoids, to pack the quaternion, which
    the current logic does not do.
(d) Fix_hmc stores this vector for each owned atom
(e) When an atom migrates (same time as neighbor list rebuild), the
    copy_arrays() and pack/unpack_exchange() methods work with the
    stored state vector for each atom
(f) The restore() method loops over currently owned atoms and
    calls a new unpack_state() method in AtomVec to restore the
    needed quantities for a single owned atom.
(g) This means restored atoms may now be owned by the wrong procs (unavoidable),
    because their new x is replaced by an old x.
    Invoke comm->exchange() and comm->borders() to migrate the atoms
    back to the original procs and populate the ghost atoms values correctly.
    Rebuild the neighbor lists (unavoidable).

(8) For rigid systems:

I think it should be the job of fix rigid/small to do the same thing
for saving/restoring rigid body info.  I.e. fix_hmc calls new
store_state() and restore_state() methods in fix rigid/small.  Its
copy_arrays() and pack/unpack_exchange() methods ensure that
additional per-atom state (which includes both per-atom and per-body
info) is handled correctly, if the "state" option is enabled.

I don't see that the FixRigidSmall class needs to be refactored.
Leave the per-body data as an internal struct.  Just add a few
new public methods.

sjplimp commented:
For rigid systems:

I think it should be the job of fix rigid/small to do the same thing
for saving/restoring rigid body info.  I.e. fix_hmc calls new
store_state() and restore_state() methods in fix rigid/small.  Its
copy_arrays() and pack/unpack_exchange() methods ensure that
additional per-atom state (which includes both per-atom and per-body
info) is handled correctly, if the "state" option is enabled.

I don't see that the FixRigidSmall class needs to be refactored.
Leave the per-body data as an internal struct.  Just add a few
new public methods.

sjplimp commented:
(9) Rigid body interface:

I don't think any other LAMMPS command which works with rigid bodies,
"depends" on the RIGID package, and includes fix_rigid_small.h.  I'm
not clear why that needs to be done by fix_hmc.  Some combination of
invoking methods in FixRigidSmall, or using extract() to query values
seems like it should be possible.  E.g. a new method in fix
rigid/small for restting vcm and omega/angmom for each body for a
chosen temperature.  Methods similar to this already exist I think,
e.g. for thermostatting.

Also, for higher-accuracy NVE, fix rigid/nve/small should be used,
not fix rigid/small.

sjplimp commented:
(10) Logic for pe/atom and press/atom and save/restore in fix hmc:

I'm not following why this is needed, including the reverse comm.  Is
the purpose to restore the energy of the system after MC rejection,
in case thermo output is done at the end of the last (rejected) MD step?
If so, maybe the eng/virial values should just be set to 0.0, to indicate there is no valid eng/virial on that timestep.
Or the timestep could be reset to the earlier save step (triggering no output).
Or an effective "run 0" could be triggered as part of stage (7g) above, to recalculate
all the energies/virials with the restored state.

sjplimp commented:
(11) New fix hmc:

Following these suggestions, I think fix hmc would become simpler:
only the end_of_step() method and the simplified save/restore state
methods suggested in (7g).  Plus copy() and pack/unpack() exchange.
Let other LAMMPS classes/methods do the work for some of the
operations now coded directly in current fix hmc.

No direct access to data in the Atom class or FixRigidSmall data.

sjplimp commented:
(12) HMC as a fix versus command:

The little I do know about HMC, I would have thought a more natural
way to implement it might be an "hmc" command rather than a fix.
Simliar to the REPLICA/hyper command (actually not a multi-replica
method), invoking a series of short MD runs and handling the
accept/reject/restore logic, which in its case involves event detection and quenches.

I think this would allow more flexibility in resetting the timestep
counter or altering the HMC algorithm.  I.e. doing regular MD between
trial MD runs, or optional extra traditional MC trial moves, followed
by MD relaxation, etc.

But if all you want to do is essentially a long continuous MD run with
trial accept/rejects/restores once each Nevery steps then I think the
fix hmc approach may work equally well.

Just something for you to think about, esp if you envision future enhancements to HMC in LAMMPS.

akohlmey commented:
@draneyj it would be nice to have some feedback from you on the various comments made by @sjplimp. Thanks.

draneyj commented:
Hi, sorry for my radio silence. I've repeatedly fallen ill and have been sick for almost a month straight now. Other serious events impacted my life and productivity around the time I got the notification for these comments. I'm finally getting back on my feet now. I promise this is a high priority for me, and I won't flake on you guys. I do really want to see this through.

RE: the comments by @sjplimp

(1) I'll pass this feedback along to the person who wrote the docs page and work with them to try to improve things a bit. Regarding the limitations of HMC, I tried to write the fix so someone could use it however they wanted, even if they did something silly like added a fix deposit, assuming they could figure out the theory behind the moves. That being said, I don't touch the internal state of any thermostats or anything, so I'll need to add a warning on the docs for that and maybe add a printed warning to the class when they're used.

(2) I originally had the timestep for the example scripts quite large, creating unstable configurations and therefore about a ~50% rejection rate (which is desirable), and think we should change it back. Also, I changed the output of the fix so the sample logs need to be updated as well.

(3) I'll revisit the input and setup methods. I think I misunderstood their purpose.

(4) Yes I agree the fix shouldn't define its own integrator. I'll update it accordingly.

(5) I am pretty sure the timestep should not be reset. When you reject a move in MC, you re-count the old configuration to get proper statistics.

(6) The discussion regarding limiting nevery revolved around a proposed solution to a problem caused by another bug I eventually found and fixed. I removed any limitations to nevery.

(7) The current logic is definitely messy. I agree that a cleaner solution would be better. I originally wanted to minimize how much I changed existing classes. In any case, I need to stare at the AtomVec class for a little bit before I can comment fully on the proposal here, but I like it so far. I'm just worried about the complexity of adding a new pack_state method for every atom style. A long shot, but could you envision us consuming the pack/unpack restart methods?

(8) I'll work on making fix rigid small pull its own weight.

(9) I'll look into this. I agree it would be great to remove the dependency.

(10) Mainly the pe/atom and press/atom stuff in fix hmc was included because 1) it was in the original code from many years ago I was using as a basis, and 2) I'm not familiar enough with lammps etc. to know if these things are important at any time between end_of_step when a configuration would be restored and when they're recalculated on the next step. I think adding an effective run 0, though maybe a little inefficient, is the best option for this as there is a relevant and real energy/virial at this point (that of the restored system), and the "MC time" has indeed moved forward. Any advice on how to do this/where else it has been done would be appreciated.

(12) I don't personally have a use case for HMC, so I can't envision it's future or how command vs fix may benefit or detract from that. If you feel strongly about this, then I'd be ok with making the change to a command. That being said, a command would be much much more difficult for me, personally, to implement, so I'd prefer to stick with the fix.

draneyj committed:
added checks for seed, temp ge 0.0

draneyj commented:
@sjplimp my other hesitations with the proposed changes in comment (7) are 1) exchanging stored data is sort of a waste of computation since, if a move is accepted that data just gets overwritten, and if a move gets rejected those atoms are now on the wrong processor, and 2) it doesn't allow for atoms to be added or removed from the system, meaning this fix can't be used in combination with e.g. gcmc. I'm not sure anyone would want to do that, but it would be counterproductive to remove that functionality if we don't need to. If we keep and restore all the data per-processor as the code currently does, everyone ends up on their original processor. Then we can just invoke comm->exchange and comm->borders to account for any processor domains which may have changed.

As for also storing and restoring quaternions, etc. In theory, the current code should do this, since it stores and restores _all_ the data in atom->peratom. That is, if I understood the structs correctly. I haven't tested it.

draneyj committed:
added flags for single execution to init and setup

draneyj committed:
pass integrator as arg for rigid, typo fix

sjplimp commented:
Replying to @draneyj comments:

>>> (5) I am pretty sure the timestep should not be reset. When you reject a move in MC, you re-count the old configuration to get proper statistics.

Not sure we are talking about the same thing.  Assume the length of trial moves is 100 steps.  Then say 4 trials are accept, reject, accept, reject.  Should the system now be at timestep 200 or 400?

>>> (7) my other hesitations with the proposed changes in comment (7) are 1) exchanging stored data is sort of a waste of computation since, if a move is accepted that data just gets overwritten, and if a move gets rejected those atoms are now on the wrong processor, and 2) it doesn't allow for atoms to be added or removed from the system, meaning this fix can't be used in combination with e.g. gcmc. I'm not sure anyone would want to do that, but it would be counterproductive to remove that functionality if we don't need to. If we keep and restore all the data per-processor as the code currently does, everyone ends up on their original processor. Then we can just invoke comm->exchange and comm->borders to account for any processor domains which may have changed. As for also storing and restoring quaternions, etc. In theory, the current code should do this, since it stores and restores all the data in atom->peratom. That is, if I understood the structs correctly. I haven't tested it.

The main reason I don't think the current implementation of save/restore state in fix_hmc.cpp is not robust is I think it fails when a reneighbor occurs during a trial run.  This is because the info in atom->peratom (vector of PerAtom entities) is not the complete "state" of each atom. (1) the quaternions of ellipsoids are not stored in PerAtom.  They are stored in a Bonus data struct within the AtomVecEllipsoid class.  Ditto for other atom styles with Bonus data.  (2) Lots of fixes store per-atom data and provide methods to migrate it with atoms when a reneighbor takes place (grep for pack_exchange() methods in fixes).  That data is not accessible to PerAtom.  If a reneighbor occurs during a trial run which is rejected, then you are restoring the coords, vels, and other PerAtom properties.  But you are not restoring the peratom data stored in those fixes.  So in other words, those restored atoms will have incorrect fix data (b/c that info changed during the reneighbors of the trial run).

I need to think some more whether the solution is what I proposed in (7) or to somehow augment what you are doing with this additional info.

draneyj commented:
> Not sure we are talking about the same thing. Assume the length of trial moves is 100 steps. Then say 4 trials are accept, reject, accept, reject. Should the system now be at timestep 200 or 400?

It should be 400

> I need to think some more whether the solution is what I proposed in (7) or to somehow augment what you are doing with this additional info.

I'm happy to do whatever you think is best. I defer completely to your judgement on this. Could I just use `(un)pack_exchange` from `AtomVec` in the same way `Irregular::migrate_atoms` does, but loop over local atoms? I could exchange the stored buffer too. I would have to add a block to keep `fix_hmc` from running its own exchange methods in the middle of storing, but that's easy. Then I don't have to worry about trying to do everything myself, I can just steal everyone's `pack_exchange` functionality.

In the mean time, I'll keep working on your other suggestions.

draneyj committed:
fix negation operator and arg starting number

akohlmey commented:
@draneyj any updates on this pull request?

draneyj commented:
I'm sorry, there aren't for now. It is still on my list of things to work on, but I've been absolutely throttled this semester. Hopefully I can make some progress soon. Thank you for checking in.

akohlmey commented:
@draneyj no worries. The purpose of asking is to determine which features will be ready for the next stable release. Based on your feedback, I will now set this pull request to "Draft" status and you can then click on the "Ready for Review" button to signal it is ready for another look.

draneyj commented:
Will do. Thank you!

draneyj committed:
implement exchange store and restore

draneyj committed:
forgot to hit save

draneyj committed:
remove commented code in store

draneyj committed:
working with exchange

draneyj committed:
cleanup

draneyj committed:
change ra keyword to resample

draneyj committed:
use a larger timestep with fix rigid example

draneyj committed:
update documentation

draneyj committed:
update documentation

draneyj committed:
add exchange and borders on restored state

draneyj committed:
remove exchange and borders on restored state. Causes atoms to be lost.

akohlmey commented:
@draneyj can you give us please an update on the status of this pull request? It has been sitting idle for several months now.
If it would not be ready soon, would you mind closing it and then resubmitting when it is closer to completion?

draneyj commented:
Hi Axel,

I'll work on it this weekend. I believe I've addressed most of the concerns brought up by you guys. I changed how atomic states are stored/restored so it coopts the exchange pack and unpack system instead of trying to keep track of everything myself. This should automatically include all the important information about the atoms. I've also done a bunch of the little things you requested.

All that's left is to move the momentum randomization into fix rigid, but I've been too lazy to actually sit down and do it. Hopefully I can get that done Saturday or next week sometime.

Best,
Jack

draneyj committed:
moved body momentum randomization into fix rigid small

draneyj committed:
removed friend class status for fixHMC

draneyj commented:
@sjplimp I've created a method in fix rigid small that randomizes velocities but I don't see how this removes the need to include fix_rigid_small.h, since I still need to have it to be able to access this method. That is, unless I add the method to fix.h as well. Some guidance here would be appreciated.

akohlmey commented:
> but I don't see how this removes the need to include fix_rigid_small.h, since I still need to have it to be able to access this method

You can look at what is done with compute rigid/local
You *do* need to include the fix rigid/small header, but only in the .cpp file. For the .h file you can use a forward declaration. to store a pointer to the fix.

Inside the ::init() function, you can then do something like this:
```
  // set fixrigid

  auto ifix = modify->get_fix_by_id(idrigid);
  if (!ifix) error->all(FLERR,"FixRigidSmall ID {} for compute rigid/local does not exist", idrigid);
  fixrigid = dynamic_cast<FixRigidSmall *>(ifix);
  if (!fixrigid)
    error->all(FLERR,"Fix ID {} for compute rigid/local does not point to fix rigid/small", idrigid);
```

If the function (e.g. myfunction()) added to fix rigid/small is in the public: section you can then just call fixrigid->myfunction() without having to have friend status or more.

Alternately, you can skip the dynamic cast and just get the generic fix pointer and then add the new function to *your* fix and at the beginning of the function use `fix->extract()` to get all the information from fix rigid/small. For that some entries need to be added for the FixRigidSmall::extract() function and probably also the declaration of Body struct made public (but *not* moved outside the FixRigidSmall class, which is a bad idea).

draneyj committed:
Merge branch 'develop' into develop

draneyj committed:
moved body struct inside class

draneyj committed:
improved error handling for user specifying rigid fix

draneyj commented:
@akohlmey Since I don't really need any information out of fix rigid small I can forego the extraction. Everything else you mentioned is now done. The only change now made to fix_rigid_small.* (vs the develop branch) is the addition of the resample_momenta function. Sounds like this is minimal impact. May I press the "ready for review" button or do you think more should be done? I could use the compiler flags to remove rigid functionality when the rigid package isn't enabled for example.

akohlmey commented:
> May I press the "ready for review" button or do you think more should be done?

That choice is up to you. After the request is flagged as "ready for review" people will take a closer look (again).

>  I could use the compiler flags to remove rigid functionality when the rigid package isn't enabled for example.

I don't understand what you are asking here. When the RIGID package is not enabled, either the CMake scripting in cmake/Modules/Packages/MC.cmake or the src/MC/Install.sh in combination with src/Depend.sh will make sure that fix hmc is not compiled when the RIGID package is missing and added when the RIGID package is added.

draneyj committed:
add whitespace to end of file

akohlmey commented:
@sjplimp @athomps this seems to be waiting for a review for quite some time already. Can you please post an update or add somebody else that would be qualified to review?

akohlmey committed:
Merge branch 'develop' into develop

akohlmey committed:
Update MC.cmake

add missing `endif()`

akohlmey committed:
Update MC.cmake

add missing `set_property()` call

sjplimp commented:
@akohlmey ok - I will look this over again this week

sjplimp committed:
Merge branch 'develop' into hmc

sjplimp commented:
@draneyj I'm looking at this again to see if it is ready to merge.  Are you still available to answer Qs or advise on any needed changes?

sjplimp committed:
reformatted code and doc page

draneyj commented:
@sjplimp yes I'm available

sjplimp commented:
One conceptual Q for the doc page.  I thought of HMC (h = hybrid) as an algorithm where you make a trial MC "move", then you relax the system for N steps before calculating the energy change due to the move.  Then accept/reject via the Metropolis criterion.  Rinse and repeat.  A "move" could be changing the position of a molecule within a dense system, or it could be inserting a new molecule into a dense system.  Normal MC would always reject these kinds of "moves" due to a large energy penalty.  But if you do MD first it might relax enough to allow the insertion via Metropolis.  This style of HMC requires soft enough potentials that you can do MD on a highly overlapped system.

But I think this fix is not doing any MC "moves" of that kind?  It is simply running MD for a long time, and every N steps it is comparing the energy at the start and end of the N-step window to either accept or reject the last N steps?  If reject, then it restores the system to the state N steps ago, resets random velocities, and continues on?

If so, that's fine.  But I'd like to add a few sentences to the doc page to explain what this accomplishes.  I.e. what is the end-state of the HMC simulation? Is it somehow an efficient way to minimize the energy of a complex system where simpler kinds of displacement "moves" in traditional MC would not do the job?

draneyj commented:
From what I understand, hmc or hamiltonian monte carlo does the latter, that is runs MD at a large, non-energy conserving time step and then accepts or rejects the new configuration as you described. The advantage is it creates a semi-reasonable proposed move for all the atoms at once. It samples the nvt distribution in this way.

draneyj commented:
From Frenkel and Smit, if you are curious:

A technique that has achieved much attention during the past few years
is the hybrid Monte Carlo scheme [430]. The basic idea behind this scheme
is that one can use MD to generate Monte Carlo trial moves. At first sight,
there is no advantage in doing so. However, the criteria for what constitutes
a good Monte Carlo trial move are more tolerant than the specifications of
a good Molecular Dynamics time step. In particular, one can take a time
step that is too long for MD. Energy will not be conserved in such a trial
move. However, as long as one uses an algorithm that is time reversible and
area preserving (i.e., that conserves volume in phase space), such collective
moves can be used as a Monte Carlo trial move. Fortunately, a systematic
way now exists to construct time-reversible, area-preservingMD algorithms,
using the multiple-time-step MD scheme of Tuckerman et al. [71]. The usual
Metropolis algorithm can then be used to decide on the acceptance or
rejection of the move (see, e.g., [431, 432]). For every trial move, the
particle velocities are chosen at random from a Maxwell distribution. In fact, it is
often advantageous to construct a trial move that consists of a sequence of
MD steps. The reason is that, due to the randomization of the velocities, the
diffusion constant of the system becomes quite low if the velocities are randomized
well before the natural decay of the velocity autocorrelation function.

sjplimp committed:
changes to doc page and cosmetic changes to src files

sjplimp committed:
whitespace changes

sjplimp commented:
Hi @draneyj and @akohlmey -

I pushed some cosmetic code edits and re-worked the doc page somewhat.

For @draneyj, here is what I think is still needed:

(a) Read the doc page and see if it all looks good.  I think some of
the command syntax had changed versus the source code, so I updated it
to what I think is correct.

(b) The doc page needs refs in the text to the 4 citations at the
bottom.  They could all be in the 1st paragraph, or maybe add a few
sentences near the top which indicate what the reader could learn from
each of the citations (or how they relate to this implementation in
LAMMPS).  Note my explanation of how you can run a hybrid rigid/flexible
system (flexible = solvent) and see if you think this is correct.

(c) Source code - not clear what the line comm_forward = 12 is for,
since this fix does not perform any forward comm - so probably not
needed.

(d) The 2 example scripts no longer work.  Possibly this is b/c I
introduced bug(s) in the source code, but I hope not.  Possibly this
is b/c those scripts are old and do not match the current required
syntax.  E.g. there is no fix nve or fix rigid/small command in them.
So I changed the syntax to what I think is now correct (see the
in.hmc*new scripts).  The flexible one now runs but does not give the
same answer as the old one.  The rigid one throws a PPPM error after
100 steps.  But the script w/out the fix hmc command throws a PPPM
error as well.  So possibly this means this data file is not well
initialized to run with rigid bodies?  Also note, that the old
log files (2023 date) are still in examples/mc.  The 4 new ones for
this version have a 2025 date.

For @akohlmey - I think fix_hmc.cpp should verify that a fix
rigid/small and/or fix nve commmand exists, depending on the "rigid"
keyword (see the doc page).  It could also be fix rigid/nve/small I
think.  Or any of their accelerated variants.  Some new code is needed
for that.

sjplimp commented:
I should add that once this is all addressed - I think this PR is ready to merge !

sjplimp committed:
spelling edits

akohlmey committed:
replace outdated inputs with new versions

akohlmey committed:
store fix rigid/small id and check if it still exists during init()

akohlmey committed:
modernize and improve error messages

akohlmey committed:
minor tweaks and modernizations, apply clang-format

akohlmey commented:
@draneyj I made some small updates and modernizations. I've removed the duplicate example and log files so we don't forget to remove them. You can always retrieve the old versions from git, if needed.

draneyj committed:
track KE when momenta aren't resampled

draneyj committed:
ensure fix_rigid images are correct after false exchange

draneyj committed:
fix examples

draneyj committed:
remove old log files

draneyj committed:
remove erroneous comm_forward = 12

draneyj committed:
fixed documentation

draneyj committed:
no longer need to check for tags

draneyj committed:
randomize solvent velocities as well when fix rigid enabled

draneyj commented:
@sjplimp I updated the documentation so it is clear and matches the code. I changed out the references since my project partners decided to originally cite a bunch of nonsense it seems. I've also fixed the examples, as well as the code itself so it functions properly within those examples... Some hairy oversights on my part were the cause of some of the errors, now corrected.

I've also removed the check for atom tags, which aren't needed (as far as I know) for the "exchange" method of restoring atom state that I've switched to. Let me know if anything else needs to be done and I'll try to get it done swiftly.

draneyj commented:
Also, new log files for the examples need to be produced, which I'm unsure how to do.

sjplimp committed:
cosmetic tweaks and generate log files

sjplimp commented:
@draneyj Everything looks good - I think you are correct, the check for atom tags is no longer needed.  I made a handful of cosmetic changes, then generated the log files for the examples.  

The only remaining issue is the the 4-proc rigid example throws an error when the HMC starts after 100 steps (step 600) with the rigid body info being messed up.  I tried setting the HMC N to 100 instead of 200 and same error.  I tried running with a 1.0 timestep (instead of 5.0) and it runs longer but similar error.  Tried running on 2 procs, similar error.  My guess is that some attributes of rigid bodies are not being restored correctly in parallel via the exchange mechanism.  I can look into that, but thought you may have seen this before already, or have ideas?

sjplimp commented:
Just noticed both the the 1-proc rigid and flexible examples are giving valgrind errors (likewise 4-proc runs).  So I suggest you fix those issues first, then the rigid body errors may go away as well.

Commit:
fix valgrind complaints

Commit:
add explicit comm exchange

Commit:
add domain calls

Commit:
fix example

Commit:
memory management matching comm classes

draneyj commented:
Hi all,

These mistakes are embarrassing. I believe I have solved the issues causing lost atoms in the examples, but cannot figure out how to avoid valgrind issues in calls to grow_store. I feel like I'm doing everything right considering I'm basically replicating the code in the comm_xxx classes for creating and growing the buffers. 

It seems I get leaks even if I just memory->create a large enough buffer in setup.

akohlmey committed:
plug memory leaks

draneyj commented:
oh duh 🤦

akohlmey commented:
> I feel like I'm doing everything right considering I'm basically replicating the code in the comm_xxx classes for creating and growing the buffers.

There were two issues:
- you didn't destroy the `buf_store` array in the destructor
- you forcibly initialized `buf_store` to nul without checking if it was already initialized (as it happens, if you have multiple "run" commands in your input).

In the first case you have a "still reachable" buffer, in the second case it is "definitely lost".

You are not the first and won't be the last person that overlooks these kinds of details.

I've developed the following strategy to minimize the chances and to make the leaks easier to detect:
- initialize all pointers to nul in the constructor initializer
- always "destroy before create" or "delete before new"
- make sure all buffers are destroyed (even if the may already have been deleted, "destroy" is safe to use on nul pointers).

Also, it makes not sense to call `memory->destroy()` before `memory->grow()`.

The inputs are not leaking anymore for me.

I see "funny" jumps in the temperature by a factor of over 10 for the rigid input deck, though.

akohlmey committed:
whitespace

sjplimp committed:
update example log files

sjplimp commented:
I updated all 4 of the log files and pushed them.

Axel wrote:
>> I see "funny" jumps in the temperature by a factor of over 10 for the rigid input deck, though.

@draneyj I assume this is b/c of the large timestep and is "ok" in the HMC context?  So long as the time integration conserves energy.  If so, would probably be good to add a line in the fix hmc doc page to the NOTE about the timestep.  To say that a large timestep may cause unphysical particle overlaps and large temperature perturbations but is OK so long as total energy (pe + ke) is conserved.

I will approve this for merging - and let Axel proceed when this final Q is resolved.

Thanks for polishing everything up @draneyj and pushing this PR across the finish line.

draneyj commented:
In fact total energy change is what the metropolis condition is based on with HMC. Not conserving total energy is the whole point. Along with this will come large changes in temperature (especially near the beginning of of a run) etc. Users will have to balance errors from large overlaps causing e.g. lost atoms during runs, acceptance ratio, timestep size, and time between "samples". I can add a paragraph to the documentation describing this clearly.

Commit:
clarify documentation

sjplimp commented:
I had forgotten the Metropolis criterion here was total energy.  I went back to read the Frenkel/Smit book quote you included above and am now somewhat confused again.  If HMC *relies* on non-energy conservation, then does it mean if you choose a small (reasonable) MD timestep for HMC and the NVE velocity-Verlet in LAMMPS does a "perfect" job of conserving energy, then the HMC simulation will be incorrect (or inefficient or pointless) ?  F/S mention rRESPA and not VV, but I would have presumed it was fine to use either (in LAMMPS) for running HMC.  If your added paragraph to the doc page will clarify this, that would be great.

akohlmey commented:
> Users will have to balance errors from large overlaps causing e.g. lost atoms during runs, acceptance ratio, timestep size, and time between "samples".

Have you experimented with the "soft core" potentials from the FEP package?
Even if you soften the potentials only a little bit, the extremely high forces and resulting disruptions from close contacts are much reduced.

draneyj commented:
I mention a couple of times in the docs page that a large timestep is required for a good MC simulation. A smaller timestep would lead to a relatively inefficient (and pointless) exploration of phase space, with most moves accepted. In the limit of a "perfect" timestep you just get MD back.

As for rRESPA vs VV, I think both are OK. IIRC someone somewhere has done the math showing it's fine to use either. But I could be wrong.

As for soft core potentials, I have not experimented with them. Sounds cool. I don't actually use HMC in my work and chose this as a project just because it was requested and sounded interesting so I'm not well versed in the tricks people use to make it easier. This also means I might not be able to write the perfect docs page for the method that includes all the minutia e.g. differences between using rRESPA vs VV.

Do the edits I made in my last commit cover enough of your questions or should I do some more research and / or clarify things a little more?

I appreciate your attention to detail on these things as I want to do a good job.

akohlmey committed:
Merge branch 'develop' into draneyj/develop

akohlmey committed:
whitespace

sjplimp commented:
@draneyj I think the paragraph on the timestep choice -- which I contributed to as I recall based on my incorrect understanding :) -- needs to be stronger.  I.e. you must use a larger-than-MD timestep (it currently doesn't say required) so that energy is *not* conserved (counter-intuitive from an MD perspective), otherwise HMC will not work efficiently (will just be the same as MD).  But if the timestep is too big, something bad will happen (what?).  Aim for an acceptance ratio of ?? percent which is  evidence that HMC is working correctly.

That's the kind of guidance I would be looking for as a user.

akohlmey commented:
Out of curiosity, I just swapped the pair style in the rigid example. Here are the changed lines:
```patch
--- in.hmc.rigid	2025-07-03 18:57:31.576672356 -0400
+++ in.hmc.soft	2025-07-03 19:03:58.806190188 -0400
@@ -13,12 +13,13 @@
 neighbor        2.0 bin
 neigh_modify    every 1 delay 0 check yes
 
-pair_style      lj/cut/coul/long 9.0
+pair_style      lj/cut/coul/long/soft 2 0.5 10.0  9.0
 pair_modify     tail yes mix arithmetic
 kspace_style    pppm 1.0e-5
 
-pair_coeff      1 1 0.1554015 3.166
-pair_coeff      2 2 0 0
+variable lambda equal 0.95
+pair_coeff      1 1 0.1554015 3.166 ${lambda}
+pair_coeff      2 2 0.0 1.0 ${lambda}
 
 bond_style      harmonic
 bond_coeff      1 554.1349 1.000
```

sjplimp commented:
I think the softer FEP potentials might be a red herring, b/c I think they might simply allow a larger timestep which stilll maintains "perfect" energy conservation.

sjplimp commented:
@draneyj Sorry, I wasn't reading the latest version of your doc page updates.  The first paragraph mentioning the timestep is good. But the NOTE paragraph dilutes it somewhat (remove Typically ?).  You could also add to it a couple of the guidelines-for-users ideas I included in my last comment to you above.  Thanks for iterating on this with me!

sjplimp commented:
You could also add (if it's true) that seeing large temperature fluctuations within each N step window is a "good sign" that HMC is working correctly (not perfectly conserving energy) ?

Commit:
add further guidance to documentation

draneyj commented:
OK I think the updated documentation now covers all the bases. Let me know what you think.

sjplimp commented:
excellent - thanks @draneyj 
@akohlmey - this is ready to merge form my perspective

akohlmey committed:
spelling and add missing reference declarations to citations

akohlmey committed:
improve errors. add warning if non-nve time integration is used w/o rigid

akohlmey committed:
move constants from header to implementation, replace defines, and remove unused ones

akohlmey committed:
tweak message to remain compatible with derived classes calling the same function
